app/src/main/java/nu/yona/app/ui/LaunchActivity.java

```diff
@@ -213,14 +213,40 @@ private void setListeners()
 
 	private void validateYonaPasswordEncryption()
 	{
+		EncryptionMethod encryptionMethod = EncryptionMethod.values()[getSharedUserPreferences().getInt(YONA_ENCRYPTION_METHOD, EncryptionMethod.INITIAL_METHOD.ordinal())];
 		// if App is older version and user is already logged in, upgrade the encryption.
-		if ((getSharedUserPreferences().getInt(YONA_ENCRYPTION_METHOD, EncryptionMethod.INITIAL_METHOD.ordinal()) == EncryptionMethod.INITIAL_METHOD.ordinal()
-				&& !TextUtils.isEmpty(getSharedUserPreferences().getString(PreferenceConstant.YONA_PASSCODE, ""))))
+		if (!TextUtils.isEmpty(getSharedUserPreferences().getString(PreferenceConstant.YONA_PASSCODE, "")) && (encryptionMethod != EncryptionMethod.ENHANCED_BASED_ON_ANDROID_KEYSTORE))
 		{
-			YonaApplication.getEventChangeManager().getSharedPreference().upgradeYonaPasswordEncryption();
+			upgradePasswordEncryption(encryptionMethod);
 		}
+		setPasswordEncryptionModeToLatest();
+	}
+
+	private void upgradePasswordEncryption(EncryptionMethod encryptionMethod)
+	{
+		switch (encryptionMethod)
+		{
+			case INITIAL_METHOD:
+			{
+				YonaApplication.getEventChangeManager().getSharedPreference().upgradeFromInitialPasswordEncryption();
+				break;
+			}
+			case ENHANCED_STILL_BASED_ON_SERIAL:
+			{
+				YonaApplication.getEventChangeManager().getSharedPreference().upgradeFromSerialBasedPasswordEncryption();
+				break;
+			}
+			default:
+			{
+				break;
```

Comment:
Please throw an exception here (e.g. ``throw new YonaRuntimeException("Unknown encryption method: " + encryptionMethod);``) and add a case label for ``ENHANCED_BASED_ON_ANDROID_KEYSTORE`` with just a ``break`` and a comment explaining that this is the current version.

That way, you will get a clear exception of you ever extend the enum and forget to update this method.